### DISCUSSION TITLE
MVUX - How to refresh a ListFeed

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno.extensions#2360] [Q&A]
Given a model like this, and having the Machines list bound to a FeedView, how can I refresh the list every time a filter flag changes?

```csharp
namespace OsmoManager.ViewModels;

internal partial record MachineBoardModel(MachinesService MachinesService)
{
    // filter flags
    public IState<bool> Unassigned => State.Value(this, () => MachinesService.Unassigned);
    public IState<bool> Running => State.Value(this, () => MachinesService.Running);
    public IState<bool> Stopped => State.Value(this, () => MachinesService.Stopped);
    public IState<bool> OutOfService => State.Value(this, () => MachinesService.OutOfService);

    public IListFeed<MachinesListItem> Machines => ListFeed.Async(MachinesService.GetMachinesListAsync);

    public async ValueTask RefreshMachinesCommand(CancellationToken ct)
    {
        MachinesService.Unassigned = await Unassigned;
        MachinesService.Running = await Running;
        MachinesService.Stopped = await Stopped;
        MachinesService.OutOfService = await OutOfService;

        // refresh the list by calling MachinesService.GetMachinesListAsync ???
    }
}
```

```csharp
internal class MachinesService
{
    public bool Unassigned { get; set; } = true;
    public bool Running { get; set; } = true;
    public bool Stopped { get; set; } = true;
    public bool OutOfService { get; set; } = true;

    private IImmutableList<MachinesListItem> _machines = ImmutableList<MachinesListItem>.Empty;

    public async ValueTask<IImmutableList<MachinesListItem>> GetMachinesListAsync(CancellationToken ct)
    {
        _machines = await GetFakeMachinesListAsync(ct);

        int states = 0;
        if (Unassigned) states |= (int)MachineStates.Unassigned;
        if (Running) states |= (int)MachineStates.Running;
        if (Stopped) states |= (int)MachineStates.Stopped;
        if (OutOfService) states |= (int)MachineStates.OutOfService;

        return _machines.Where(x => ((int)x.State & states) > 0).ToImmutableList();
    }
}
```

**Comment by dr1rrb** [comment]
My suggestion here would be to change the architecture into a more declarative/reactive approach (I'm adding doc about this here https://github.com/unoplatform/uno.extensions/pull/2361)

To apply it to your case:

```csharp
namespace OsmoManager.ViewModels;

internal partial record MachineBoardModel(MachinesService MachinesService)
{
    public IState<MarchineCriteria> Criteria => State.Value(this, () => new MarchineCriteria());
    public IListFeed<MachinesListItem> Machines => Criteria.Select(MachinesService.GetMachinesListAsync).AsListFeed();
}
```

```csharp
internal partial record MarchineCriteria(bool Unassigned, bool Running, bool Stopped, out OutOfService);

internal class MachinesService
{
    private IImmutableList<MachinesListItem> _machines = ImmutableList<MachinesListItem>.Empty;

    public async ValueTask<IImmutableList<MachinesListItem>> GetMachinesListAsync(MarchineCriteria criteria, CancellationToken ct)
    {
        _machines = await GetFakeMachinesListAsync(ct);

        int states = 0;
        if (criteria.Unassigned) states |= (int)MachineStates.Unassigned;
        if (criteria.Running) states |= (int)MachineStates.Running;
        if (criteria.Stopped) states |= (int)MachineStates.Stopped;
        if (criteria.OutOfService) states |= (int)MachineStates.OutOfService;

        return _machines.Where(x => ((int)x.State & states) > 0).ToImmutableList();
    }
}
```

And you can then edit the `Criteria` from the view by simply binding 2-way to `Criteria.Unassigned`, `Criteria.Running`, etc. Each time a criteria is changed, a request will be sent and the `Machines` will be updated.

And if you still want to let the user to refresh the list, you can also use the `Refresh` command exposed by the `FeedView`. It can be used directly [within the `ListView` (e.g. pull-to-refresh)](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Mvux/Tutorials/HowTo-ListFeed.html#data-bind-the-view) or [externally (e.g. in a command bar)](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Mvux/FeedView.html#refresh-command-property)

**Reply by mcNets** [reply]
@dr1rrb: Thanks David, this clarifies things a lot. I'm not really used to this reactive behavior. By any chance can you suggest me any readings about it other than the specific doc you are adding?

**Reply by francoistanguay** [reply]
A lot of reactive style programming is inspired by how RX.Net works: https://github.com/dotnet/reactive

When looking at updating State ,the way I look at the different strategies are:

InputState (modified by User like Search Term or Filters)
OutputState (derived from an Async Query + InputState: Like a GetXYZFromCriteria or SearchFromTerm)
Optional User Explicit Action like a PullToRefresh View/Container, a Refresh button, a Search button:

Strategy 1: Every time Input State Changes, update OutpuState
  This is typically performed with the Select (Projection) operator
  IListFeed<Output> Output = Input.SelectAsync(input => Service.GetOutputAsync(input));
  
Strategy 2: Ever time a user explicitly requests a refresh, get the current Input, call the Method to get the Output

  IListState<Output> Output;
  
  void Refresh(Input input) // BY convention, if there's a Feed named Input, the Refresh method will get a snapshot of the feed's current value.
  {
    var output = await.GetOutputAsync(intput)
    
    Output.UpdateAsync(_ => output);
   }

**Reply by mcNets** [reply]
Thank's @francoistanguay.

**Reply by mcNets** [reply]
@francoistanguay 

Just one more thing, I'm getting this error:

```
CS0411	The type arguments for method 'Feed.Select<TSource, TResult>(IFeed<TSource>, Func<TSource, TResult>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
```

I tried with: 

```csharp
Criteria.Select<MachineCriteria, ImmutableList<MachinesListItem>>(MachinesService.GetMachinesListAsync);

Criteria.Select<MachineCriteria, IListFeed<MachinesListItem>>(MachinesService.GetMachinesListAsync);
```

But it's waiting for a value-type.

What would be the proper syntax?

**Reply by francoistanguay** [reply]
I think you need to use .Select(...).AsListFeed()

https://platform.uno/docs/articles/external/uno.extensions/doc/Reference/Reactive/listfeed.html#aslistfeed

**Reply by mcNets** [reply]
That's what I am using, have a look at David example:

```csharp
internal partial record MachineBoardModel(MachinesService MachinesService)
{
    public IState<MarchineCriteria> Criteria => State.Value(this, () => new MarchineCriteria());
    public IListFeed<MachinesListItem> Machines => Criteria.Select(MachinesService.GetMachinesListAsync).AsListFeed();
}
```

**Reply by francoistanguay** [reply]
Sorry, what is the question? Is the syntax above compiling and working or not? Error?

**Reply by mcNets** [reply]
Sorry @francoistanguay 

The error I'm getting here:

`public IListFeed<MachinesListItem> Machines => Criteria.Select(MachinesService.GetMachinesListAsync).AsListFeed();`

> CS0411	The type arguments for method 'Feed.Select<TSource, TResult>(IFeed<TSource>, Func<TSource, TResult>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.

**Reply by mcNets** [reply]
I'm able to get the list in this way:
```csharp
public IListFeed<MachinesListItem> Machines
{
    get
    {
        var feed = Feed.Async(
            async ct =>
            {
                var criteria = await Criteria;
                var machines = await MachinesService.GetMachinesListAsync(criteria, ct);
                return machines;
            });
        return feed.AsListFeed();
        //return Criteria.Select(MachinesService.GetMachinesListAsync).AsListFeed();
    }
}
```

**Comment by mcNets** [comment]
@francoistanguay @dr1rrb Solved, Select must be SelectAsync

**Reply by dr1rrb** [reply]
Oh sorry indeed.
